### DISCUSSION TITLE
Feature proposal: support for `composer://` stream-wrapper

### DISCUSSION THREAD

**Discussion by mindplay-dk** [composer/composer#9668] [General]
*"How do I locate files in composer packages?"*

This is a common question on any project that wasn't built using one of the large, mainstream frameworks - which offer concepts like "bundles" etc. to answer this question in various ways.

I personally build all my projects using an IOC container and hand-picked composer packages. I don't want a full framework.

Some people just hard-code the base-paths - and answer you tend to find on StackOverflow - but this creates problems, for one, the path is obviously different when the package is the root package (when you're running tests, etc.) and when it's installed in the `vendor` folder. Composer installer plugins can alter package paths, `composer.json` can change the `vendor` folder name, etc.

There are packages available to deal with this problem as well - on the light end, my own package [mindplay/composer-locator](https://packagist.org/packages/mindplay/composer-locator) accounts for these problems; and on the heavy end (now dead?) packages like [Puli](https://github.com/puli) basically built their own virtual file-systems to replace the native file-system.

Now, every template engine, markdown renderer, really anything that accepts a file path, generally uses `file_get_contents` or `fopen` or other standard functions - and so, they already support PHP-style URIs, and thereby support stream-wrappers.

So I can make this work - by choosing a package (`mindplay/composer-locator` or something like it) I can find the physical path, and then pass that path to library X.

But I'm introducing another dependency, at the source-code level, just to locate the file. This is all information that Composer *has* - packages like `mindplay/composer-locator` really just take the information from Composer and exposes it. I already *have* composer, I *have* the file installed, but now I need to pick a library and write code just to locate the file...

What I'd really like is to be able to just do this:

```php
$readme = file_get_contents("composer://vendor/package/path/to/readme.txt");
```

The `composer://` stream-wrapper would resolve the base path `vendor/package` to the installed package root-folder, and the rest of the path is simply appended to that physical path.

So I went ahead and [added that](https://github.com/mindplay-dk/composer-locator/pull/9) to `mindplay/composer-locator`.

It passes all of the following tests, where `ok` asserts `true`, and `eq` is a strict equality assertion:

https://github.com/mindplay-dk/composer-locator/pull/9/files#diff-0e9f7e9fab2837fcff42dfc571491aeae61f1090a8c6351b93a49c30f0927e21

So this very simply virtualizes the logical Composer package hierarchy into a logical file-system - you can open files using standard file-system functions, and you can use `readdir` or `DirectoryIterator` etc. to iterate over the root `composer://` folder, which will give you vendor-names, you can iterate over e.g. `composer://mindplay` to find packages from a specific vendor, etc.

It "just works" with any old PHP library, and it's all pretty intuitive, I think.

To the point where I'm wondering why this feature isn't just built into Composer. 🤔

I know it's always been a policy to never introduce actual APIs, and thereby source-code dependencies, in Composer - I've asked in the past for an API to locate package paths and gotten a solid "no", and I understand why.

But our projects do depend on Composer for the generated autoloader - it's not a source-code level dependency, but it is a dependency, as otherwise nothing would be able to load, right?

If somebody wanted to create a competing loader, they could - it would just have to parse standard JSON and support standard PSR-4 paths, etc.

This doesn't seem that different.

Granted, there would be things to consider and discuss first. For example, is `composer://` an appropriate name? It might not be - something like `packages://` might be more neutral territory? And would the `packages` scheme need to be a PSR standard?

I think it's worth discussing? 🙂

**Comment by mindplay-dk** [comment]
I should mention the one function that *doesn't* work, which is `glob` - per the [manual entry](https://www.php.net/glob):

> Note: This function will not work on remote files as the file to be examined must be accessible via the server's filesystem.

In other words, it does not support *any* type of stream-wrapper. (and custom stream-wrappers cannot support it.)

**Comment by stof** [comment]
I think such stream wrapper will be complex enough to justify being maintained as a dedicated library rather than being generated code for which we cannot easily report bug fixes or even security fixes.

And if it is a dedicated library, it does not even need to be maintainer by the composer core team then

**Reply by mindplay-dk** [reply]
The amount of code is nothing substantial, in the area of a few hundred LOC.

I could let you peep the working branch, if you're interested?

It's all pretty simple, which is why I'm wondering why this isn't just built in.

**Comment by stof** [comment]
note that such stream wrapper might leverage the `InstalledVersions` API of Composer 2 to know the list of packages

**Comment by mindplay-dk** [comment]
The code is now available to view [here](https://github.com/mindplay-dk/composer-locator/pull/9).

It's ~200 LOC for the stream-wrapper, on top of the ~100 LOC or so for the underlying library, really nothing substantial.

I do wish you would consider this as a standard Composer feature - it delivers the autoloading, but doesn't provide any means of locating your installed resources. As a result, people currently do "dumb things" (reflection, wrong assumptions about `__DIR__` etc.) or invent frameworks or packages to solve the problem.

Note that I'm *not* suggesting the class/API generated by this library be included. I'm only maintaining it for backwards compatibility in the first place - it could be completely removed (refactored away) or could be  considered just an internal implementation detail.

So this wouldn't require code-generation like my package does - a simple dump of the package names/paths to a file would do just fine.

Being able to `fopen` or `file_get_contents` out-of-the-box just seems like a perfectly natural expectation to me. Why not?

**Comment by Seldaek** [comment]
IMO https://github.com/composer/composer/issues/9648 is a better way to handle this as it gives you the base path without overhead of stream wrappers. And then you can do whatever you want with that path yourself.

**Reply by mindplay-dk** [reply]
That issue seems to be about the plugin API?

This library deals with locating installed package folders *in your project code*, at run-time.

(It is itself a plugin, which generates a file during installation, so it can't be used in a plugin either.)

**Reply by Seldaek** [reply]
Nope, it's about adding a little more information to https://getcomposer.org/doc/07-runtime.md#installed-versions - which is available both at runtime for the project and install time for plugins.

**Reply by mindplay-dk** [reply]
Wait, so now we have the API you [told](https://github.com/composer/composer/issues/5214#issuecomment-211842936) me we wouldn't have? 🙂

That's great! I wasn't even understanding what @stof was suggesting, since my understanding was that Composer would never have any sort of run-time API.

I suppose I could just retire the package entirely now.

Or I could do a new package with just the stream-wrapper, I suppose. Although now that there's a standardized way to locate the package roots, the only reason for it to exist would be convenience. Probably not really worth it.

Well, this was a great way to waste a couple of days. 😂

Oh, by the way, `getInstalledPackages()` and `getRootPackage()` are both missing from the documentation.

**Reply by Seldaek** [reply]
Yes I told you that in 2016 though, then with Composer 2 we had issues with ocramius/package-versions and decided to take this into core.

And yeah those are not listed because IMO they are likely to bite people in the ass when combined with replace/provide and misunderstandings. That's why the docs are more use-case focused to hopefully guide people to the right way to do what they want to do.

Anyway in Composer 2.1, with #9648, we'd add an install path API to this and that should really take care of what you were looking for.

**Reply by mindplay-dk** [reply]
Okay, so a little too early to retire my package - but once 9648 is implemented then, that's great! 🙂

(It does seem a little odd to leave `public` API facilities undocumented? Use-case focused or not, you made them public - documentation probably should just point out the caveats. Most people won't think to look for this documentation unless they have a good reason to anyhow...)
